File: backend/schemas/municipality.py
"""Municipality-related Pydantic schemas for Premium Dashboards."""

from pydantic import BaseModel, ConfigDict, Field
from .companies import CompanyBase

class TrendPoint(BaseModel):
    """A single data point in a time-series trend."""
    label: str  # e.g., "Jan 23"
    value: float | int

class SectorStat(BaseModel):
    """Industry sector statistics for a municipality."""
    nace_division: str
    nace_name: str
    company_count: int
    total_employees: int | None = None
    percentage_of_total: float | None = None

class RankingInfo(BaseModel):
    """Rank position relative to total entities in group."""
    rank: int
    out_of: int

class MunicipalityPremiumResponse(BaseModel):
    """Consolidated premium response for a municipality dashboard."""
    code: str
    name: str
    county_code: str
    county_name: str
    
    # Coordinates for mapping
    lat: float | None = None
    lng: float | None = None
    
    # SSB Population Data
    population: int
    population_growth_1y: float | None = Field(None, description="Percentage growth last year")
    
    # Business Metrics
    company_count: int
    business_density: float | None = Field(None, description="Companies per 1000 residents")
    business_density_national_avg: float | None = None
    
    # Performance Stats
    total_revenue: float | None = None
    avg_profit_margin: float | None = None
    
    # Trends
    establishment_trend: list[TrendPoint] = []
    
    # Categorical Data
    top_sectors: list[SectorStat] = []
    top_companies: list[CompanyBase] = []
    newest_companies: list[CompanyBase] = []
    latest_bankruptcies: list[CompanyBase] = []
    
    # Rankings
    ranking_in_county_density: RankingInfo | None = None
    ranking_in_county_revenue: RankingInfo | None = None
    
    model_config = ConfigDict(from_attributes=True)

class MunicipalityListResponse(BaseModel):
    """Lightweight response for sitemaps and lists."""
    code: str
    name: str
    slug: str
    company_count: int
    population: int | None = None
    lat: float | None = None
    lng: float | None = None

    model_config = ConfigDict(from_attributes=True)

File: frontend/src/types/municipality.ts
import { Company } from './company'

export interface TrendPoint {
    label: string
    value: number
}

export interface SectorStat {
    nace_division: string
    nace_name: string
    company_count: number
    total_employees?: number
    percentage_of_total?: number
}

export interface RankingInfo {
    rank: number
    out_of: number
}

export interface MunicipalityPremiumResponse {
    code: string
    name: string
    county_code: string
    county_name: string

    // Coordinates
    lat?: number
    lng?: number

    // SSB Population Data
    population: number
    population_growth_1y?: number

    // Business Metrics
    company_count: number
    business_density?: number
    business_density_national_avg?: number

    // Performance Stats
    total_revenue?: number
    avg_profit_margin?: number

    // Trends
    establishment_trend: TrendPoint[]

    // Categorical Data
    top_sectors: SectorStat[]
    top_companies: Company[]
    newest_companies: Company[]
    latest_bankruptcies: Company[]

    // Rankings
    ranking_in_county_density?: RankingInfo
    ranking_in_county_revenue?: RankingInfo
}

export interface MunicipalityListItem {
    code: string
    name: string
    slug: string
    company_count: number
    population?: number
    lat?: number
    lng?: number
}
